### ISSUE TITLE
Non keep-alive HTTP/1.x requests close the connection immediately with compressed responses

### ISSUE THREAD

**Issue by leobispo** [eclipse/vert.x#2184]
Hi all,

First of all, Thank you for the amazing work with vertx.

I am testing the Version 3.5.0 with an nginx reverse proxy.

If I turn off the keep-alive functionality, the HttpServerResponseImpl is not answering anything.

You can test using the following HTTP requests:

# Not working

```
GET / HTTP/1.0.
Host: my-server.com
X-Real-IP: 127.0.0.1.
X-Forwarded-For: 127.0.0.1.
Connection: close.
Cache-Control: max-age=0.
Upgrade-Insecure-Requests: 1.
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.79 Safari/537.36.
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8.
Accept-Encoding: gzip, deflate.
Accept-Language: pt-BR,pt;q=0.8,en-US;q=0.6,en;q=0.4.
.
```

# Working

```
GET / HTTP/1.1.
Host: my-server.com
X-Real-IP: 127.0.0.1.
X-Forwarded-For: 127.0.0.1.
Cache-Control: max-age=0.
Upgrade-Insecure-Requests: 1.
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.79 Safari/537.36.
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8.
Accept-Encoding: gzip, deflate.
Accept-Language: pt-BR,pt;q=0.8,en-US;q=0.6,en;q=0.4.
.
```

This is because of the `Connection: close`

After a debug, I found that if HttpServerResponseImpl line 442 keepAlive is false, it is working as expected

```JAVA
    if (!keepAlive) {
      closeConnAfterWrite();
      closed = true;
    }
```

So I suspect that the closeConnAfterWrite is not working as expected.

I would appreciate if you can take a look on it.

**Comment by vietj** [comment]
thanks we will investigate soon

**Comment by vietj** [comment]
can you provide a reproducer ? I've got a case that works here.

**Comment by vietj** [comment]
with HTTP/1.0 the `Connection: close` header has no meaning, what matters is the `Connection: keep-alive` header.

**Comment by leobispo** [comment]
The easiest way to reproduce is to create a nginx reverse proxy to a vertx http app and then try to access it via the proxy

**Comment by vietj** [comment]
can you quickly provide an nginx configuration so we can try it ?

**Comment by vietj** [comment]
or can you provide a wireshark dump of the socket between nginx/vertx ?

**Comment by leobispo** [comment]
The following configuration should do the trick.

```
server {
  listen       *:80;
  server_name  my-server.com;

  access_log            /var/log/nginx/my-server-access.log combined;
  error_log             /var/log/nginx/my-server-error.log;

  location / {            
    proxy_pass            http://localhost:8080;
    proxy_read_timeout    90;
    proxy_connect_timeout 90;
    proxy_set_header      Host $host;
    proxy_set_header      X-Real-IP $remote_addr;
    proxy_set_header      X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header      Proxy "";
  }
}
```

**Comment by vietj** [comment]
can you give a CURL command line to send to the proxy ?

**Comment by leobispo** [comment]
This will call the proxy. Nginx will redirect to you app on port 8080

curl -XGET http://my-server.com

**Comment by leobispo** [comment]
Don't forget to add the my-server to your host pointing to 127.0.0.1

**Comment by vietj** [comment]
I can't see where HTTP/1.0 is configured

**Comment by leobispo** [comment]
it is default... if you don't pass in the configuration

proxy_http_version    1.1;
proxy_set_header      Connection "";

nginx will assume 1.0

**Comment by vietj** [comment]
ok, thanks will investigate

**Comment by vietj** [comment]
I've just tried and the server disconnects after the response is sent, here is my server app test:

```
    Vertx vertx = Vertx.vertx();
    HttpServer server = vertx.createHttpServer();
    server.requestHandler(req -> {
      System.out.println(req.version());
      req.response().end("Hello World");
    }).listen(8081);
```

full nginx config:

```
events {
}
http {
  server {
    listen       *:8080;
    server_name  my-server.com;
    location / {
      proxy_pass            http://localhost:8081;
      proxy_read_timeout    90;
      proxy_connect_timeout 90;
      proxy_set_header      Host $host;
      proxy_set_header      X-Real-IP $remote_addr;
      proxy_set_header      X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_set_header      Proxy "";
    }
  }
}
```

**Comment by leobispo** [comment]
This example worked for me as well...

**Comment by vietj** [comment]
maybe it comes from the application ?

**Comment by leobispo** [comment]
I don't know. It looks like that the connection is closed before the event loop flushes the data. Maybe because your example is too simple, or there is nothing else in the event loop, things seems to work.

But this is just my guess, because when it is keep-alive it is working flawless

**Comment by leobispo** [comment]
Ok, just to give you one more feedback. Even with keep-alive, after sometime the app started to not respond connections.

With 3.4.2 it is working perfectly

**Comment by vietj** [comment]
I will add a delay to see how it goes

**Comment by vietj** [comment]
it would be great if you can investigate and give more insight about what is causing the issue

**Comment by leobispo** [comment]
I will.

**Comment by vietj** [comment]
can you try with 3.5.0.Beta1 also ?

**Comment by leobispo** [comment]
Yes... Right now I can't, but I will take tomorrow to help you to investigate this issue.

**Comment by vietj** [comment]
sure, I'll monitor this issue

**Comment by vietj** [comment]
is your server using event loop or perhaps you are using worker or different thread to end the request ?

**Comment by vietj** [comment]
can you tell what you observe on the connection from the server to the proxy ? it's not clear what is happening.

**Comment by leobispo** [comment]
Just to give you a feedback. I am trying to debug the code, but it is a bit hard to follow the logic... specially because of the event loop async code.

I couldn't find the reason for the connection close... but it looks like there are still remaining data on the channel after it gets closed. 

The Netty code handler is checking if the remaining event is already cancelled, but that is not the case.

I just found one more issue in my app. My client app is doing 3 calls when it is logged. The first call is OK, but the next 2 are stuck. I checked the code and it looks like the .next() is never calling the next route.

I don't know if the 2 bugs are related.

Sorry if I cannot help you in more details.

Please let me know if there are someone else facing the same problems as I am. Also let me know if I can help you to figure out the 2 issues.

**Comment by leobispo** [comment]
The second bug I found. You changed a bit the way how the Auth is done. Because I have a custom Auth is broke my logic.

It remains the first one. But it happens only when Connection: close

**Comment by vietj** [comment]
so far nobody has reported issues, it would be good to separate the two issues and handle them separately.

**Comment by vietj** [comment]
we need to handle in this issue, only the bug related to the behavior of `HttpServer` you observe in your case

**Comment by vietj** [comment]
@leobispo any news about this issue ?

**Comment by leobispo** [comment]
@vietj No news. I changed the nginx to send the keep-alive header and the problem is not happening.

So, we can assume that the problem is something related to the connection being closed before the end of the flush.

I will try to create a small app that simulates the problem tomorrow. I think it will be easier for debugging.

But I am starting to suspect that the error is in the netty.

**Comment by vietj** [comment]
@leobispo ok, closing this for now, reopen if you find something

**Comment by mikand13** [comment]
Im seeing similar behaviour to this. All our microservices are behind nginx reverse proxies. Been running fine on 3.5.0.Beta1 but started seeing issues on 3.5.0. Specifically event threads start throwing this:

[vert.x-eventloop-thread-2] ConnectionBase - java.nio.channels.ClosedChannelException

All application side handlers are processed as expected by vertx-web, but the nginx "client" reports this:

2017/10/26 13:30:18 [error] 8#8: *13 upstream prematurely closed connection while reading response header from upstream, client: 127.0.0.1, server: localhost, request: "GET /api/elb-status HTTP/1.1", upstream: "http://10.2.0.5:8080/api/control/ping", host: "localhost"

So the connection is closed while reading the response, throwing a 502 to the original client.

Is resolved by adding 

proxy_http_version 1.1;

to the configuration for the elb path in nginx. (Our Healthcheck)
Normal traffic is fine as its upgraded for SockJS anyway.

**Comment by vietj** [comment]
@mikand13 can you by chance investigate this ?

**Comment by mikand13** [comment]
@vietj Can do, will let you know what I figure out.

**Comment by vietj** [comment]
can you try to investigate without vertx web ?

**Comment by mikand13** [comment]
It would appear that the HTTP version was a decoy. The actual culprit is setting compressionsupport on the httpserver. I havent investigated exactly why but i have created an example you can run to see the different nginx responses.

https://github.com/mikand13/vertx-conn-close-test

mvn clean package && docker-compose up

/one/elb-status and /onlyZip/elb-status will give 502. /elb-status gives 200.

Talking directly to the vertx application works fine, so there is something going on in the interaction with nginx.

**Comment by vietj** [comment]
thanks, will have a look today

**Comment by vietj** [comment]
so you suspect it's between vertx response and nginx ?

**Comment by mikand13** [comment]
Correct, works fine when f.ex. curling directly to the Vertx app. In all three alternatives.

**Comment by vietj** [comment]
pulling docker images :-)

**Comment by vietj** [comment]
it's not clear which curl command I should use and what I should observe

**Comment by vietj** [comment]
so 

```
curl -v http://localhost:5001/elb-status
```

according to the docker compose file

**Comment by mikand13** [comment]
Correct. Usually works a couple of attempts, then breaks.

**Comment by mikand13** [comment]
Should run /one/elb-status or /onlyZip/elb-status tho for the compression enabled endpoints in the Vertx app.

**Comment by vietj** [comment]
I can reproduce without docker and using my own ngninx install with a couple of modifs to the nginx conf on hosts / include

**Comment by vietj** [comment]
indeed it's related to the connection: close header sent by the client with a compressed response that produce an empty response, that should help to create at test case and investigate quickly

**Comment by mikand13** [comment]
Ill throw in an example explanation for anyone else looking at this later. This is docker output:

![docker](https://user-images.githubusercontent.com/6736611/32092268-fd7c2a4c-baf7-11e7-813a-d9d6abb19567.png)

Healthchecks on the nginx are running. You will observe that /elb-status gives a 200 and both /one/elb-status (HTTP/1.1) and /one/onlyZip, which are the two compression enabled endpoints return 502, as well as the vertx app throwing exceptions.

**Comment by vietj** [comment]
I see internally this happening when the connection close header is sent

```
io.netty.handler.codec.EncoderException: java.lang.IllegalStateException: unexpected message type: io.netty.handler.codec.http.DefaultHttpContent (expected: HttpResponse)
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:106)
	at io.netty.handler.codec.MessageToMessageCodec.write(MessageToMessageCodec.java:116)
	at io.vertx.core.http.impl.HttpChunkContentCompressor.write(HttpChunkContentCompressor.java:36)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:816)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:723)
	at io.netty.handler.stream.ChunkedWriteHandler.doFlush(ChunkedWriteHandler.java:304)
	at io.netty.handler.stream.ChunkedWriteHandler.flush(ChunkedWriteHandler.java:137)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush0(AbstractChannelHandlerContext.java:776)
	at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:768)
	at io.netty.channel.AbstractChannelHandlerContext.flush(AbstractChannelHandlerContext.java:749)
	at io.vertx.core.net.impl.ConnectionBase.endReadAndFlush(ConnectionBase.java:88)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:342)
	at io.vertx.core.impl.ContextImpl.executeFromIO(ContextImpl.java:200)
	at io.vertx.core.net.impl.VertxHandler.channelReadComplete(VertxHandler.java:140)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:398)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:380)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelReadComplete(AbstractChannelHandlerContext.java:373)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelReadComplete(ChannelInboundHandlerAdapter.java:97)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:398)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:380)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelReadComplete(AbstractChannelHandlerContext.java:373)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelReadComplete(ChannelInboundHandlerAdapter.java:97)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:398)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:380)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelReadComplete(AbstractChannelHandlerContext.java:373)
	at io.netty.handler.codec.ByteToMessageDecoder.channelReadComplete(ByteToMessageDecoder.java:324)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:398)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:380)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelReadComplete(AbstractChannelHandlerContext.java:373)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelReadComplete(DefaultChannelPipeline.java:1364)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:398)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelReadComplete(AbstractChannelHandlerContext.java:380)
	at io.netty.channel.DefaultChannelPipeline.fireChannelReadComplete(DefaultChannelPipeline.java:941)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:139)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:645)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:580)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:497)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:459)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: unexpected message type: io.netty.handler.codec.http.DefaultHttpContent (expected: HttpResponse)
	at io.netty.handler.codec.http.HttpContentEncoder.ensureHeaders(HttpContentEncoder.java:240)
	at io.netty.handler.codec.http.HttpContentEncoder.encode(HttpContentEncoder.java:96)
	at io.netty.handler.codec.http.HttpContentEncoder.encode(HttpContentEncoder.java:51)
	at io.netty.handler.codec.MessageToMessageCodec$1.encode(MessageToMessageCodec.java:67)
	at io.netty.handler.codec.MessageToMessageEncoder.write(MessageToMessageEncoder.java:88)
	... 41 more
```

**Comment by vietj** [comment]
it's also related to the fact that the client does not set an `accept-encoding` header

**Comment by vietj** [comment]
I've pushed a fix in vertx-core master for this case, I've tested with @mikand13 reproducer and it works fine. Can you check on your side ?

**Comment by mikand13** [comment]
Can do

**Comment by mikand13** [comment]
Just tested now, and it works fine, thanks! Will close the test repo.

**Comment by draxly** [comment]
Found this bug report after many hours of trying to find a solution myself, happy to see it addressed already. 
My situation (and perhaps someone has a similar setup) was that the health checks for AWS Target Groups (for their Application Load Balancers) as well as Consul health checks stopped working for my HTTP server, which had compression enabled, after updated to Vertx 3.5.0. I had to do some workaround to get this running, therefore eagerly awaiting 3.5.1. =)
